### DISCUSSION TITLE
How does Kademlia caching _work_?

### DISCUSSION THREAD

**Discussion by joshuef** [libp2p/rust-libp2p#4561] [Q&A] [answered]
So, I've looked to enable `KademliaCaching` in the kad config. As I understand it anythign `Quorum::One` should automatically be cached at the closest X nodes to the data...

Now, I'm not really sure how to verify this is happening, or where it happens. I have poked about the code, but didn't turn up anything.

So I'm wondering:

- Is caching automatic?
- Are records cached _separately_ to the record store? (I'd assumed so, but perhaps not?)
- Can the cache size be controlled?  And if so, where?
- Can we identify/filter/act on caches records at nodes?

Thanks for any pointers here! 🙇

**Comment by mxinden** [comment, marked as answer]
> Is caching automatic?

It has been in the past with `Quorum::One`, but is no longer. We have removed the `Quorum` mechanism from `get_record`. See https://github.com/libp2p/rust-libp2p/pull/2712. The user needs to manually call `Kademlia::put_record_to` on a `GetRecordOk::FinishedWithNoAdditionalRecord`.

Note that the current docs are outdated, as they still reference the `Quorum` on the `get_record` code path. Would you mind creating a pull request updating the docs @joshuef?

https://github.com/libp2p/rust-libp2p/blob/38ea7ad4534565876225454a71edf91ce67adb16/protocols/kad/src/behaviour.rs#L2778-L2789

> Are records cached separately to the record store? (I'd assumed so, but perhaps not?)

I am not sure I understand the question. Maybe the following helps. The local node looks up a key. It discovers multiple peers that should but don't actually have the key. When the user now calls `put_record_to` with the key, value and the peers, the local node will attempt to put the record to the peer just as it would with the normal `Kademlia::put_record`. Just that this time it knows upfront which peers to target. From the peers perspective, it can not differentiate the write-back caching from a normal `put_record`.

> Can the cache size be controlled? And if so, where?

In case you are using the `MemoryStore` you can use the `MemoryStoreConfig::max_records` option.

> Can we identify/filter/act on caches records at nodes?

No. Remote notes are ignorant of the fact that an incoming put record is actually from a cache write-back.

I am sorry that this is not more intuitive. Hope this helps @joshuef.

**Reply by thomaseizinger** [reply]
Does the term caching come from the paper @mxinden? If not, we could rename this to make it clearer.

**Reply by joshuef** [reply]
Ah yeh that helps a lot, thanks @mxinden 🙇 

 I'll try and whip up some docs as per my new understanding and PR that in.

So `cache_candidates` really is just "these folk should/could have the data" as per the caching config. But it all needs to be done manually via `put_record`.

What I'm still missing from the picture here: given `GetRecord` queries _should_ go only to the data holders, how and when is cache envisaged to be utilised? Alongside GetProvider as a backup mechanism?

**Reply by mxinden** [reply]
> Does the term caching come from the paper @mxinden? If not, we could rename this to make it clearer.

Yes, this wording is taken from the paper:

> For caching purposes, once a lookup suceeds, the requesting node stores the <key, value> pair at the closest node it observed to the key that did not return the value.

https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf

---

> So cache_candidates really is just "these folk should/could have the data" as per the caching config. But it all needs to be done manually via put_record.

Via `put_record_to`, not `put_record`!

> What I'm still missing from the picture here: given GetRecord queries should go only to the data holders, how and when is cache envisaged to be utilised? Alongside GetProvider as a backup mechanism?

One iteratively sends `GetRecord` requests to the 20 closest nodes, hoping that eventually a node returns the value for a given key. A peer might know the value because the original author of the key value pair send a `PutRecord` request to it. Or a peer might know the value because some other peer, after having looked it up, send a `PutRecord` request as part of its caching follow-up.

Note that none of this involves **Provider** records.

Does that make sense @joshuef?

**Reply by joshuef** [reply]
Yep, making sense 👍 . 

One last Q: `GetRecord` is always to the `k_value` nodes, not `ReplicationFactor` nodes _specifically_ (I'd thought it was the latter)?

ie, our replication factor is `8`, but `k_value` is 20 as per defaults. So any `GetRecord` req is to the `k_value` nodes and may not actually hit the _actual_ closest nodes?

**Reply by mxinden** [reply]
`GetRecord` iteratively queries the `k_value` closest nodes. It terminates on either:

- It found the record value.
- The `k_value` closest peers did not return the value AND they don't know anyone closer.

> ie, our replication factor is 8, but k_value is 20 as per defaults. So any GetRecord req is to the k_value nodes and may not actually hit the actual closest nodes?

This is an eventual consistent system with transient connections, thus there is no guarantee whatsoever. In an ideal world (high consistency, no network faults) a `k_value` and `replication_factor` of `1` suffices. In the real world the higher the `k_value` the more likely you will find the value. The higher the `replication_factor` the more likely for your value to be found.

Why do you use a `replication_factor` of `8`? I would stick to the defaults which are battle tested on the IPFS network, which is a very decentralized network with lots of churn.

**Reply by joshuef** [reply]
Okay grand, good to know!

> Why do you use a replication_factor of 8?

We're using a different incentive mechanism to (hopefully) enable us to lower replication counts and reduce churn (as well as some other mechanisms to reduce the weight of churn), effectively removing the need for altruism there. 

As things stablilise I'm hopeful we might be able to bring some of this upstream to you all. At the moment we're still testing and tweaking right now!


Thanks for all the insight @mxinden 🙇 . A PR will follow shortlish.

**Reply by joshuef** [reply]
Here's the PR that I think clarifies the documentation: https://github.com/libp2p/rust-libp2p/pull/4564

**Reply by mxinden** [reply]
> As things stablilise I'm hopeful we might be able to bring some of this upstream to you all. At the moment we're still testing and tweaking right now!

Maybe something to present at one of our [open maintainer calls](https://github.com/libp2p/rust-libp2p/discussions/4556) or the [libp2p community call](https://github.com/libp2p/specs/discussions/577) @joshuef?

**Reply by joshuef** [reply]
Ah sorry, missed that @mxinden ! Yeh I could talk show you how our replication flow is working at the moment and how/where that's sticking for us or perhaps counter to how libp2p might be intended.

**Reply by mxinden** [reply]
Great. Let me know if this works for you. https://github.com/libp2p/specs/discussions/588

**Reply by joshuef** [reply]
Yeh that should be fine 👍
